### PR TITLE
implement a fast atan approximation function

### PR DESCRIPTION
**Pull request by zhichen3**
implement a fast atan function using an approximate function for x in [-1, 1], for |x| > 1, the identity `arctan(x) = sign(pi/2, x) - arctan(1/x)` (Thanks to Eric) is used.

Two versions of the approximate functions are included.
1) Efficient Approximations for the Arctangent Function by Rajan 2006: Equation 9: Max error ~0.0015 rad
2) https://stackoverflow.com/questions/42537957/fast-accurate-atan-arctan-approximation-algorithm: Max error ~0.00063 rad

Generally I find the form from stackoverflow is more accurate and also faster. 1) is also discussed in 2).

In general, from testing, atanf is roughly more than 2 times faster than std::atan.

### CODE REVIEW

File: util/approx_math/fast_atan.H [hunk 1]
```diff
@@ -0,0 +1,114 @@
+#ifndef MICROPHYSICS_APPROX_MATH_H
```
Comment (zingale) [inline util/approx_math/fast_atan.H:1]: I would suggest calling the header approx_math.H
this way we can add other approximations in the future to the same header

File: util/approx_math/fast_atan.H [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    ///
+    /// This calculates atan within [-1, 1] range.
+    ///
+
+    // formula 1
+    // constexpr amrex::Real PI_4 = GCEM_PI / 4.0_rt;
+    // return PI_4 * x;
+
+    // formula 2
+    // constexpr amrex::Real A = 0.0776509570923569_rt;
```
Comment (zingale) [inline util/approx_math/fast_atan.H:23]: can you add the reference to where these expressions came from?

**Comment by zhichen3** [comment]
Not sure why hip failed to compile. Also I kept the volatile qualifier for testing because runtime wouldn't increase with iteration otherwise.

**Comment by zingale** [comment]
I suspect the HIP failure is due to changes in ROCm and there is a namespace clash now.

**Comment by zingale** [comment]
can you try not including `<AMReX_GpuUtility.H>` and instead directly include just what you need?

**Comment by zingale** [comment]
or alternately, include `<AMReX_Gpu.H>`, since that seems to do some namespace magic.

**Comment by yut23** [comment]
It's probably this line: https://github.com/AMReX-Astro/Microphysics/blob/30469dd4b2c8374b5a19708f95f6921748211366/networks/iso7/actual_network.H#L10

**Comment by yut23** [comment]
`atanf()` may conflict with the definition from the standard library. I think `fast_atan()` works fine.

**Comment by zingale** [comment]
indeed, we should get rid of those `using namespace amrex`
I'll try that in a separate PR

**Comment by zhichen3** [comment]
I figured checking nan is unnecessary, so I removed it and also removed `AMReX_GpuUtility.H` since I don't need isnan anymore. I guess this also fixed hip compilation error.

**Review by yut23 — APPROVED** [review]

**Comment by zhichen3** [comment]
sounds good to me

**Pull request merged**